backend/controller/evaljob_controller.go

```diff
@@ -0,0 +1,568 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	lmevalservicev1beta1 "github.com/foundation-model-stack/fms-lm-eval-service/api/v1beta1"
+	"github.com/go-logr/logr"
+)
+
+const (
+	DriverPath                  = "/bin/driver"
+	DestDriverPath              = "/opt/app-root/src/bin/driver"
+	PodImageKey                 = "pod-image"
+	DriverImageKey              = "driver-image"
+	DriverServiceAccountKey     = "driver-serviceaccount"
+	DefaultPodImage             = "quay.io/yhwang/lm-eval-aas-flask:test"
+	DefaultDriverImage          = "quay.io/yhwang/lm-eval-aas-driver:test"
+	DefaultDriverServiceAccount = "driver"
+)
+
+// EvalJobReconciler reconciles a EvalJob object
+type EvalJobReconciler struct {
+	client.Client
+	Scheme    *runtime.Scheme
+	Recorder  record.EventRecorder
+	ConfigMap string
+	Namespace string
+	options   ServiceOptions
+}
+
+type ServiceOptions struct {
+	PodImage             string
+	DriverImage          string
+	DriverServiceAccount string
+}
+
+// +kubebuilder:rbac:groups=lm-eval-service.github.com,resources=evaljobs,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=lm-eval-service.github.com,resources=evaljobs/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=lm-eval-service.github.com,resources=evaljobs/finalizers,verbs=update
+// +kubebuilder:rbac:groups="",resources=pods,verbs=get;list;watch;create;delete
+// +kubebuilder:rbac:groups="",resources=configmaps,verbs=get;watch;list
+
+func (r *EvalJobReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := log.FromContext(ctx)
+
+	evalJob := &lmevalservicev1beta1.EvalJob{}
+	if err := r.Get(ctx, req.NamespacedName, evalJob); err != nil {
+		log.Error(err, "unable to fetch EvalJob. could be from an deletion request")
+		// we'll ignore not-found errors, since they can't be fixed by an immediate
+		// requeue (we'll need to wait for a new notification), and we can get them
+		// on deleted requests.
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	if !evalJob.ObjectMeta.DeletionTimestamp.IsZero() {
```

Comment:
Interesting. Is this how a `finalizer` is triggered for a `kubebuilder` controller?

Reply:
yes!

Reply:
actually, for the controller, it's reconciliation fashion, which means you don't know which event triggers the `reconcile` function. and the DeletionTimestamp is the field that says the current state of the CR is being deleted. After the CR is deleted, there will be another `reconcile` call. but I use a `predicate` to ignore the deletion events since we have nothing to do after the CR has been deleted.